code/parachain/runtime/picasso/src/contracts.rs

```diff
@@ -182,36 +141,19 @@ impl PalletHook<Runtime> for Precompiles {
 		let dex: AccountIdOf<Runtime> = PabloPalletId::get().into_account_truncating();
 		match contract_address {
 			address if address == dex => {
-				let message: cw_dex_router::msg::ExecuteMsg = serde_json::from_slice(message)
-					.map_err(|_| CosmwasmVMError::ExecuteDeserialize)?;
-				match message {
-					cw_dex_router::msg::ExecuteMsg::Swap { in_asset, min_receive, pool_id } => {
-						let in_asset_id = AssetToDenom::convert(in_asset.denom)
-							.map_err(|_| CosmwasmVMError::AssetConversion)?;
-						let min_receive_asset_id = AssetToDenom::convert(min_receive.denom)
-							.map_err(|_| CosmwasmVMError::AssetConversion)?;
-						let in_asset_amount = in_asset.amount.into();
-						let min_receive_amount = min_receive.amount.into();
-						let who = AccountAddrConvert::convert(
-							vm.0.data().cosmwasm_message_info.sender.clone().into_string(),
-						)
-						.map_err(|_| CosmwasmVMError::AccountConvert)?;
-						let result = <Pablo>::do_swap(
-							&who,
-							pool_id.into(),
-							AssetAmount::new(in_asset_id, in_asset_amount),
-							AssetAmount::new(min_receive_asset_id, min_receive_amount),
-							true,
-						);
-						match result {
-							Ok(result) => {
-								let response = Response::new()
-									.add_attribute("amount", result.value.amount.to_string());
-								Ok(ContractResult::Ok(response))
-							},
-							Err(err) => Ok(ContractResult::Err(alloc::format!("{:?}", err))),
-						}
-					},
+				let message: composable_traits::dex::ExecuteMsg =
+					serde_json::from_slice(message)
+						.map_err(|_| CosmwasmVMError::ExecuteDeserialize)?;
+
+				let result = common::dex::DexPrecompile::<_, _, Pablo>::execute(
+					vm.0.data().cosmwasm_message_info.sender.as_str(),
+					message,
+				)
+				.map_err(|_| CosmwasmVMError::<Runtime>::Precompile);
+
+				match result {
+					Ok(result) => Ok(ContractResult::Ok(response)),
+					Err(err) => Ok(ContractResult::Err(alloc::format!("{:?}", err))),
 				}
 			},
 			_ => Err(CosmwasmVMError::ContractNotFound),
```

Comment:
Since you’re reflowing this perhaps simplify the control flow be
replacing the match with an `if` with an early return:

```
			if address != dex {
				return Err(CosmwasmVMError::ContractNotFound);
			}

			let message: composable_traits::dex::ExecuteMsg = ...
```

Reply:
early return is not simplificaiton. actually early returns are again what you asked later split match by functions. early returns is lie. there so you propose moving _ => case up? but in rust convetion to move down.

Reply:
early returns produce irrefactorable code and again you opinionated desires earlier.

Reply:
deeply nested code ideas are also opinionated. if you look at relation of files vs directories, you will see the real notion of nesting.